test/validator.js

```diff
@@ -0,0 +1,113 @@
+import assert from "assert";
+import path from "path";
+import child_process from "child_process";
+import {readdir, readfile, writefile, PASS, FAIL} from "./util.js";
+import * as FTL from "../syntax/ast.js";
+
+const ABSTRACT = `${__dirname}/../syntax/abstract.js`;
+const GRAMMAR = `${__dirname}/../syntax/grammar.js`;
+const FIXTURES = `${__dirname}/fixtures`;
+
+main();
+
+async function main() {
+    let changed = await modify_grammar();
+    changed |= await validate_selector();
+    if (changed) {
+        child_process.execSync("git diff syntax",  { stdio: "inherit" });
+    }
+    process.exit(Number(changed));
+}
+
+async function modify_grammar() {
+    let changed = false;
+    let grammar = await readfile(GRAMMAR);
+    let to_replace = /(not|maybe)\(/g;
+    let m, iteration = 0;
+    while (m = to_replace.exec(grammar)) {
+        let replacement, new_grammar = grammar.slice(0, m.index);
+        switch (m[1]) {
+            case "not":
+                replacement = "maybe";
+                break;
+            case "maybe":
+                replacement = "not";
+                break;
+        }
+        new_grammar += replacement;
+        new_grammar += grammar.slice(m.index + m[1].length);
+        await writefile(GRAMMAR, new_grammar);
+        console.log(`Grammar validation iteration ${++iteration}`);
+        const keep_change = verify_fixtures();
+        if (keep_change) {
+            grammar = new_grammar;
+            changed = true;
+            console.log("new grammar");
+        }
+    }
+    await writefile(GRAMMAR, grammar);
+    return changed;
+}
+
+async function validate_selector() {
```

Comment:
It would be great to have a short comment at the top of this and previous functions which in describe tl;dr of what those tests are attempting to perform and test.